### PR TITLE
Make auth managers provide their own airflow CLI commands

### LINKED ISSUE #32206: AIP-56 - FAB AM - Move User and Role CLI commands to FAB AM
User and Role Rest APIs need to be moved to FAB Auth manager

### LINKED ISSUE #32208: AIP-56 - Implement mechanism so auth managers can define additional CLI commands
Auth managers might need to define additional CLI commands. e.g. User and Role CLI commands need to be defined in the FAB auth manager. The purpose of this task is to implement such mechanism.

### LINKED ISSUE #32348: AIP-56 - Move sync-perm command
It has been brought up by Jarek in this [discussion](https://github.com/apache/airflow/discussions/32187#discussioncomment-6323078). I dont have much context on it but it seems we need to move this command as well as part of the AIP-56 work

### PR DESCRIPTION
**Pull request by vandonr-amz**
closes #32206 
closes #32208 
closes #32348

taking a lot of inspiration from #29055
Since auth managers are going to be configurable, we don't want their specific commands to be defined in airflow core.
Thus, this PR moves (almost) everything auth manager-related to the fab corner, and adds a mechanism so that auth managers can provide their own CLI commands.

On my computer, running this with the FAB auth manager makes the commands take 10 to 15 extra milliseconds, which seems reasonable.

### CODE REVIEW

File: airflow/auth/managers/fab/cli_commands/utils.py
Comment (vincbeck) [inline airflow/auth/managers/fab/cli_commands/utils.py:1]: Should we move this file in FAB? I dont think it is something specific to FAB right?

File: airflow/cli/commands/standalone_command.py [hunk 1]
```diff
@@ -182,7 +182,7 @@ def initialize_database(self):
         # server. Thus, we make a random password and store it in AIRFLOW_HOME,
         # with the reasoning that if you can read that directory, you can see
         # the database credentials anyway.
-        from airflow.utils.cli_app_builder import get_application_builder
+        from airflow.auth.managers.fab.cli_commands.utils import get_application_builder
```
Comment (vincbeck) [inline airflow/cli/commands/standalone_command.py:185]: This is not good. We do not want Airflow to have a dependency on FAB auth manager. People might use another Auth manager and not even have FAB auth manager installed (this is assuming FAB auth manager is moved to a separate provider)

File: airflow/providers/celery/executors/celery_executor.py [hunk 2]
```diff
@@ -143,6 +142,7 @@ def __getattr__(name):
 )
 
 # worker cli args
+ARG_AUTOSCALE = Arg(("-a", "--autoscale"), help="Minimum and Maximum number of worker to autoscale")
```
Comment (vincbeck) [inline airflow/providers/celery/executors/celery_executor.py:145]: Seems unrelated to the PR no?

**Review by vincbeck — CHANGES_REQUESTED** [review]

**Comment by vincbeck** [comment]
Does it resolve #32348 as well?

File: airflow/auth/managers/fab/cli_commands/utils.py
Comment (vandonr-amz) [reply airflow/auth/managers/fab/cli_commands/utils.py:1]: yes it is, it contains 2 methods, `_return_appbuilder(app: Flask) -> AirflowAppBuilder` and `get_application_builder() -> Generator[AirflowAppBuilder, None, None]`, which are very specific to FAB.

If we need to add other util methods to core CLI classes of course we can recreate a utils.py file in the original location.

File: airflow/cli/commands/standalone_command.py (hunk 1 above)
Comment (vandonr-amz) [reply airflow/cli/commands/standalone_command.py:185]: yes I know, we talked a bit about it on Thursday when we looked at the code together :)
I moved the "easy" commands in the PR, this one will need to move (at least partially) too, but the situation is a little bit more complex, I suggest dealing with it in a separate PR, where I'd introduce an `init_airflow` method in auth managers for instance.

File: airflow/providers/celery/executors/celery_executor.py (hunk 2 above)
Comment (vandonr-amz) [reply airflow/providers/celery/executors/celery_executor.py:145]: yes, somewhat, but it was defined with the `# roles` args, and is actually only used here, so I moved it to its logical home.

**Comment by vandonr-amz** [comment]
> Does it resolve #32348 as well?

ah yes indeed, I didn't see there was a separate ticket for this

File: airflow/auth/managers/fab/cli_commands/utils.py
Comment (vincbeck) [reply airflow/auth/managers/fab/cli_commands/utils.py:1]: Sorry I mis-explained. I meant, it is not related to FAB auth manager no? These two functions seems to be needed regardless of the auth manager no?

File: airflow/cli/commands/standalone_command.py (hunk 1 above)
Comment (vincbeck) [reply airflow/cli/commands/standalone_command.py:185]: I am okay with that but we definitely do not want to forget it. This is something we really dont want

File: airflow/auth/managers/fab/cli_commands/utils.py
Comment (vandonr-amz) [reply airflow/auth/managers/fab/cli_commands/utils.py:1]: hmm, you mean we should leave them as abstract methods (i.e. keep the declaration in place) and only move the implementation to the FAB folder ?
I'm not sure I agree with you, I think those methods are not needed outside of the commands I moved (or plan to move), so I don't see why we'd keep them in core 🤔

Comment (vincbeck) [reply airflow/auth/managers/fab/cli_commands/utils.py:1]: Oh I see, somehow I thought `get_application_builder()` was needed elsewhere but if it is only used for FAB auth manager related CLI command, I am good with that

File: airflow/providers/celery/executors/celery_executor.py (hunk 2 above)
Comment (o-nikolas) [reply airflow/providers/celery/executors/celery_executor.py:145]: This one felt generic enough to leave in the cli_config for other future worker based executors, but I don't feel too strongly.

File: airflow/cli/cli_parser.py [hunk 3]
```diff
@@ -62,6 +63,13 @@
     # Do no re-raise the exception since we want the CLI to still function for
     # other commands.
 
+try:
+    auth_mgr = get_auth_manager()
```
Comment (o-nikolas) [inline airflow/cli/cli_parser.py]: It looks like this helper gets the class from config and constructs an instance before returning. Could we update that to take an optional arg (or add another helper) to just return the class? For executors the `get_cli_commands` function is a static method and you just need the class imported to call it, no need to create an instance (i.e. run `__init__` and all it's side effects).

Also how expensive is `init_auth_manager` and the fab class itself to import? It's important to optimize anything on the CLI path to be very fast to import since it will slow down the CLI otherwise. You can see some examples of what I did in the executor space and the benchmarks I provided:
- Optimizing the [base executor class](https://github.com/apache/airflow/pull/30361)
- Optimizing [Celery](https://github.com/apache/airflow/pull/31001) and [Kubernetes](https://github.com/apache/airflow/pull/30727) executor classes.

Comment (vandonr-amz) [reply airflow/cli/cli_parser.py]: ah yes good point. We discussed it with Vincent, and he was in favor of a non-static method in case implementers want to use instance properties, but what you say about init time makes sense, I'll change it.

Comment (vincbeck) [reply airflow/cli/cli_parser.py]: I am good with that then

**Comment by vandonr-amz** [comment]
@vincbeck you left a "requesting changes" review, I think your concerns have been addressed ?

**Review by vincbeck — APPROVED** [review]

**Comment by vincbeck** [comment]
> @vincbeck you left a "requesting changes" review, I think your concerns have been addressed ?

Approved

File: airflow/cli/cli_parser.py (hunk 3 above)
Comment (o-nikolas) [reply airflow/cli/cli_parser.py]: @vandonr-amz Did you end up looking at the affect of the import time of the fab auth manager (second paragraph in my first comment)?

I think it's important to do at least some quick benchmarking to see how slow or fast it is. Before I optimized the executors they added seconds to the CLI runtime just because loading the module was slow.

Comment (vandonr-amz) [reply airflow/cli/cli_parser.py]: ok I will.

Comment (vandonr-amz) [reply airflow/cli/cli_parser.py]: taking a quick look, I put a timer at the top of `cli_parser.py` and another one just after the test on duplicate subcommands (at the end of top level code).
Without this code we're at ~30ms
With, it's taking ~90ms
So this costs ~60ms.

Comment (vandonr-amz) [reply airflow/cli/cli_parser.py]: I can bring the total time down to 40-45ms by getting rid of the `from flask import url_for` in fab auth manager.

**Review by o-nikolas — APPROVED** [review]
